std/encoding.d

```diff
@@ -3907,3 +3907,121 @@ if (isForwardRange!Range && is(Unqual!(ElementType!Range) == ubyte))
 
 /** Constant defining a fully decoded BOM */
 enum dchar utfBOM = 0xfeff;
+
+
+/++
+    Strips the BOM (if present) from the front of the given range and returns
+    the range.
+
+    If the range is a range of $(D ubyte), then $(LREF getBOM) will be used to
+    determined if there is a BOM present, and the matching BOM will be stripped.
+
+    If the range is a range of $(D char) or an array of $(D char), then the
+    UTF-8 BOM will be stripped if present, but no other BOMs will be checked
+    for.
+
+    If the range is a range of $(D wchar) or an array of $(D wchar), then the
+    UTF-16 BOM of the native endianness will be stripped if present, but no
+    other BOMs will be checked for.
+
+    If the range is a range of $(D dchar) or an array of $(D dchar), then the
+    UTF-32 BOM of the native endianness will be stripped if present, but no
+    other BOMs will be checked for.
+
+    Params:
+        range = A range of characters of ubyte.
+
+    Returns:
+        The range with the BOM stripped from the front of the given range if
+        there is a BOM at the front of the range; otherwise, the given range is
+        returned unchanged.
+  +/
+R stripBOM(R)(R range)
+    if (isForwardRange!R && (isSomeChar!(ElementType!R) || is(Unqual!(ElementType!R) == ubyte)))
+{
+    import std.encoding : getBOM;
+
+    alias E = Unqual!(ElementEncodingType!R);
+
+    static if (is(E == ubyte))
+    {
+        range.popFrontN(range.save.getBOM().sequence.length);
+        return range;
+    }
+    else
+    {
+        import std.algorithm.searching : startsWith;
+        import std.utf : byCodeUnit;
+
+        auto bcu = range.byCodeUnit();
+
+        static if (is(E == char))
+        {
+            import std.range : only;
+            if (bcu.save.startsWith(only(cast(char)0xEF, cast(char)0xBB, cast(char)0xBF)))
+                bcu.popFrontN(3);
+        }
+        else static if (is(E == wchar))
+        {
+            if (bcu.save.byCodeUnit().startsWith(cast(wchar)0xFEFF))
+                bcu.popFront();
+        }
+        else
+        {
+            if (range.save.byCodeUnit().startsWith(cast(dchar)0x0000FEFF))
+                bcu.popFront();
+        }
+
+        static if (isNarrowString!R)
+            return bcu.source;
+        else
+            return bcu;
+    }
+}
+
+///
+unittest
+{
+    import std.string : representation;
+
+    assert(stripBOM("hello world") == "hello world");
+
+    assert(stripBOM((cast(string)[0xEF, 0xBB, 0xBF]) ~ "hello world") ==
+           "hello world");
+
+    assert(stripBOM("\uFEFFhello world"w) == "hello world"w);
+
+    // utf-7
+    ubyte[] utf7 = (cast(ubyte[])[0x2B, 0x2F, 0x76, 0x38, 0x2D]) ~
+                   "hello world".representation;
```

Comment:
I'm not sure about this... this is going back to the previous problem of accidentally stripping a piece of UTF-8 text just because it happens to be UTF-7.

I think that the `ubyte` overload should take a `BOM` parameter, which tells it which `BOM` to strip. It should then only strip only that `BOM` (if it's present).

If the program isn't sure which BOM is in the content, it should use `getBOM` first. Stripping arbitrary BOMs doesn't make sense anyway, because if you don't know or care what BOM is in the data, then you won't know the data's encoding and how to process it correctly!

Reply:
I don't know. I honestly don't care much about encodings other than UTF-8, UTF-16, and UTF-32. Stripping off whichever BOM was there made sense to me for ubytes, since if you want the BOM gone, at that point you have to check them all, though yeah, if you want to do anything meaningful with the range at that point, you already have to know what the encoding is at that point, and you probably had to call `getBOM` anyway. I don't see much point to `stripBOM` if you have to pass it the BOM, especially if you already had to call `getBOM`. At that point, it's trivial to just pop the elements based on the return value of `getBOM`. So, I guess that I can just remove ubyte support from `stripBOM`.

Reply:
If we don't support ranges of `ubyte` though, maybe this belongs in std.utf, since it's UTF-specific.